### ISSUE TITLE
Currency not Latin fields issue

### ISSUE THREAD

**Issue by vomikan** [moneymanagerex/moneymanagerex#739]
![screenshot 29](https://cloud.githubusercontent.com/assets/6836805/13382946/6f54c20e-de91-11e5-8c1d-b20412b1b8b6.jpg)
- [x] Windows 7x64
- [x] 1.3.0-alpha

**Comment by vomikan** [comment]
~~~~~~
INSERT INTO CURRENCYFORMATS_V1 VALUES(119,'São Tomé and Príncipe dobra','Db','','.',' ','','',100,1,'STD');
~~~~~

**Comment by guanlisheng** [comment]
it works fine for mac

**Comment by vomikan** [comment]
On INSERT for all currencies the result incorrect if non ASCII symbols used
![screenshot 37](https://cloud.githubusercontent.com/assets/6836805/13668088/72c97230-e6cb-11e5-995f-2eef00ccc3e8.jpg)

**Comment by vomikan** [comment]
Also incorrect code
~~~~~
wxTRANSLATE("United States dollar"), wxTRANSLATE("$"), wxTRANSLATE(""), wxTRANSLATE("."), wxTRANSLATE(" "), wxTRANSLATE(""), wxTRANSLATE(""), wxTRANSLATE("USD")));
~~~~~
wxTRANSLATE(...) only for name needed. Should be 
~~~~~
wxTRANSLATE("United States dollar"), L"$", L"", ".", " ", L"", L"", "USD")));
~~~~~

**Comment by vomikan** [comment]
I've discovered that the code:
~~~~~
    wxString s = wxString::Format(L"INSERT INTO CURRENCYFORMATS_V1 VALUES (2, '%s', '%s', '%s', '%s', '%s', '%s', '%s', 100, 1, '%s')"
        , "European euro", L"€", "", ".", " ", "", "", "EUR");

~~~~~
Working fine if placed, for example, in maincurrencydialog.cpp
![screenshot 39](https://cloud.githubusercontent.com/assets/6836805/13697091/32161ef8-e77c-11e5-978f-1538dadc452f.jpg)

But this code in DB_Table_Currencyformats_V1.h provide another result:
![screenshot 40](https://cloud.githubusercontent.com/assets/6836805/13697267/632a2736-e77d-11e5-8ae5-77835f031a30.jpg)

**Comment by vomikan** [comment]
@stef145g could you reproduce this issue?

**Comment by stef145g** [comment]
I have now tested this and it works to some degree but this is not the solution.

I have created a temporary branch: https://github.com/stef145g/moneymanagerex/tree/test_only_non_functional
to allow others to experiment with it.

1. I have created **DB_Table_Currencyformats_V1.cpp with a few methods moved to the this file.
2. I have modified the euro currency Unicode character in the .cpp file
3. I put similar code in mmframe.cpp. Warning: can only be executed once.

When starting the program and immediately creating a new database, the new currencies will be added. I have found that initial EUR character is wrong, but the EUR1 character is correct.

Still not sure why. Just splitting DB_Table_Currencyformats into a .h and .cpp file is not the solution.

**Comment by vomikan** [comment]
FYI
http://www.w3schools.com/charsets/ref_utf_currency.asp

https://forums.wxwidgets.org/viewtopic.php?f=23&t=41968

**Comment by vomikan** [comment]
I've converted DB_Table_Currencyformats_V1.h file to UTF-8
![screenshot 46](https://cloud.githubusercontent.com/assets/6836805/13773191/1b8b4ade-eaa9-11e5-877b-e7bbd4459f76.jpg)
It started to work fine.

Now there are a lot of warnings appears:
~~~~~
1>c:\dev\mmex\src\db/DB_Table_Currencyformats_V1.h(223): warning C4566: character represented by universal-character-name '\u00E3' cannot be represented in the current code page (1251) (..\..\src\assetspanel.cpp)
~~~~~
But 
~~~~~
#pragma execution_character_set("utf-8")
~~~~~
helps to fix this warnings.

**Comment by vomikan** [comment]
Related links: 
http://www.skorkin.com/2012/07/how-to-set-the-default-encoding-for-all-files-saved-in-visual-studio/#.VufZ7OKLSM8

**Comment by guanlisheng** [comment]
it's already UTF-8...
`src/db/DB_Table_Currencyformats_V1.h:         UTF-8 Unicode C++ program text, with very long lines
`

**Comment by vomikan** [comment]
There are two UTF-8 with BOM and without

**Comment by vomikan** [comment]
How to implement this changes?
~~~~~
        db->ExecuteUpdate(wxString::Format("INSERT INTO CURRENCYFORMATS_V1 VALUES (3, '%s', '%s', '%s', '%s', '%s', '%s', '%s', 100, 1, '%s')", wxTRANSLATE("UK Pound"), wxTRANSLATE("£"), wxTRANSLATE(""), wxTRANSLATE("."), wxTRANSLATE(" "), wxTRANSLATE("Pound"), wxTRANSLATE("Pence"), wxTRANSLATE("GBP")));
to
        db->ExecuteUpdate(wxString::Format("INSERT INTO CURRENCYFORMATS_V1 VALUES (3, '%s', '%s', '%s', '%s', '%s', '%s', '%s', 100, 1, '%s')", wxTRANSLATE("UK Pound"), "\u1234", "", ".", " ", wxTRANSLATE("Pound"), wxTRANSLATE("Pence"), "GBP"));
~~~~~

**Comment by vomikan** [comment]
It's vital to note that
~~~~~
wxTRANSLATE("")
~~~~~
unacceptable to use.

**Comment by guanlisheng** [comment]
@vomikan does this issue happen on Linux version? 
it works pretty fine on Mac and i am suspecting it is a platform specific issue. @moneymanagerex/desktop-mmex-team

**Comment by vomikan** [comment]
Workaround for VS is:
![screenshot 48](https://cloud.githubusercontent.com/assets/6836805/13808599/f9c16828-eb77-11e5-9890-973942feadd2.jpg)

![screenshot 47](https://cloud.githubusercontent.com/assets/6836805/13808606/ff4c0780-eb77-11e5-9182-992f2035b773.jpg)

**Comment by vomikan** [comment]
Final call here:
The file should be converted to UTF-8 with signature;
Pragma should be added:
~~~~~
#pragma execution_character_set("utf-8")
~~~~~
L prefix should be added to all non Latin strings
~~~~~
        db->ExecuteUpdate(wxString::Format("INSERT INTO CURRENCYFORMATS_V1 VALUES (4, '%s', '%s', '%s', '%s', '%s', '%s', '%s', 100, 1, '%s')", wxTRANSLATE(L"Русский рубль"), wxTRANSLATE(""), L"р", ",", " ", wxTRANSLATE(L"руб."), wxTRANSLATE(L"коп."), "RUB"));
~~~~~

**Comment by vomikan** [comment]
Personally, I believe that all non Latin symbols should go away from source files.

**Comment by vomikan** [comment]
Related fixes: https://github.com/moneymanagerex/database/pull/13/commits/618281a59ff6379ab3b62ddb79b7b516c42663a3

**Comment by guanlisheng** [comment]
would we close it right now?

**Comment by vomikan** [comment]
It's still a mess

**Comment by vomikan** [comment]
@guanlisheng please remove all entries of wxTRANSLATE("") from py file then update the code and po files. There are many helpless string present now (about 500 like "." ","  "$U" "Céntimo" "دج" etc).

**Comment by guanlisheng** [comment]
how is everything going after some wxTRANSLATE removal?

**Comment by vomikan** [comment]
![screenshot 77](https://cloud.githubusercontent.com/assets/6836805/18309262/2a6c694c-7501-11e6-8383-2c3e83479628.jpg)
![screenshot 78](https://cloud.githubusercontent.com/assets/6836805/18309352/9794279e-7501-11e6-94b8-be4f3642629b.jpg)

**Comment by vomikan** [comment]
Fix already there
https://github.com/moneymanagerex/database/pull/13/files

**Comment by guanlisheng** [comment]
what's your code version and did you try the latest release one? as the comment said, i cannot reproduce the issue and i didn't see the complaint from other ppl either.

**Comment by vomikan** [comment]
This issue is present in last release. I think it's Windows only issue.

**Comment by guanlisheng** [comment]
What the last release is? Did you try 1.3.0-xx?

**Comment by vomikan** [comment]
I am working with latest 1.3.0

пн, 12 сент. 2016 г. в 18:21, Lisheng <notifications@github.com>:

**Comment by vomikan** [comment]
http://forum.moneymanagerex.org/viewtopic.php?f=23&t=6897

**Comment by vomikan** [comment]
Working file
[DB_Table_Currencyformats_V1.h.txt](https://github.com/moneymanagerex/moneymanagerex/files/687910/DB_Table_Currencyformats_V1.h.txt)

**Comment by vomikan** [comment]
There is a source of CURRENCYFORMATS_V1 table inserts:
~~~~~~
INSERT INTO CURRENCYFORMATS_V1 VALUES(1,'United States dollar','$','','.',' ','','',100,1,'USD');
INSERT INTO CURRENCYFORMATS_V1 VALUES(2,'European euro','€','','.',' ','','',100,1,'EUR');
INSERT INTO CURRENCYFORMATS_V1 VALUES(3,'UK Pound','£','','.',' ','Pound','Pence',100,1,'GBP');
INSERT INTO CURRENCYFORMATS_V1 VALUES(4,'Russian Ruble','','₽',',',' ','руб.','коп.',100,1,'RUB');
~~~~~~
The Python script wrap all latin values into wxTRANSLATE.
Result:
~~~~~~
db->ExecuteUpdate(wxString::Format("INSERT INTO CURRENCYFORMATS_V1 VALUES ('1', '%s', '$', '', '.', ' ', '', '', '100', '1', '%s')", wxTRANSLATE("United States dollar"), wxTRANSLATE("USD")));
~~~~~~
Problem: Nobody needs translations of ISO codes and currency names as well and nobody will translate so huge amount of strings.
also
~~~~~~
db->ExecuteUpdate(wxString::Format("INSERT INTO INFOTABLE_V1 VALUES ('1', '%s', '%s')", wxTRANSLATE("DATAVERSION"), wxTRANSLATE("3")));
~~~~~~
This technology is OK only for Categories!
~~~~~~
        db->ExecuteUpdate(wxString::Format("INSERT INTO CATEGORY_V1 VALUES ('1', '%s')", wxTRANSLATE("Bills")));
        db->ExecuteUpdate(wxString::Format("INSERT INTO CATEGORY_V1 VALUES ('2', '%s')", wxTRANSLATE("Food")));
~~~~~~

**Comment by vomikan** [comment]
I suggest mark field that to be translated special symbols like 
~~~~~~
/*_*/ or /*wxTRANSLATE*/
~~~~~~

~~~~~~
INSERT INTO CATEGORY_V1 VALUES(1, /*_*/'Bills');
INSERT INTO CATEGORY_V1 VALUES(2, /*_*/'Food');
~~~~~~

Accordingly, the Python script should be changed .

**Comment by stef145g** [comment]
Currency values are initialised when the table is created. While we have a problem with the initialisation, this does not fix existing or old databases. The only way to fix existing databases is to rewrite the existing data in the currency lookup table.

The problem with doing this is that in some databases, the existing currencies will be replaced and these may not correctly align with the new currency definitions created on initialisation/readjustment.

The only way around this is to discover what files are affected, make the change then manually readjust the currency for each account that has been realigned with an incorrect currency, including the base currency.

MMEX version 1.3.x has a new command **Database Debug** under *Tools / Database*
Upgrade Check: [currency_table_upgrade_check.mmdbg.txt](https://github.com/moneymanagerex/moneymanagerex/files/692818/currency_table_upgrade_check.mmdbg.txt)
Upgrade Patch: [currency_table_upgrade_patch.mmdbg.txt](https://github.com/moneymanagerex/moneymanagerex/files/692819/currency_table_upgrade_patch.mmdbg.txt)

These two files have been provided to fix this problem as follows:
1. Down and remove the .txt extension to allow MMEX to access them.
2. Run and save the output of the upgrade_check file
3. Run the upgrade_patch file
4. Restart MMEX
5. Using the output of the upgrade_check, manually correct the currencies for
   a. The base currency
   b. Each account with the incorrect currency setting

**Comment by vomikan** [comment]
The problem in currencies for new DB (created exactly in v 1.3.x)
Your recommendation is not enough.

**Comment by stef145g** [comment]
I have used this procedure to fix my working database that was created at Version 0.9.6.0
My database still had 11 currencies initialised, and I am using the latest version of MMEX.

While this is not the total solution, it is a valid solution for all existing databases created prior to 1.3.0

**Comment by vomikan** [comment]
How to reproduce the issue:
Start mmex - create new DB - create 'Test' account with Euro currency.
Result:
![screenshot 353](https://cloud.githubusercontent.com/assets/6836805/21760129/bc847b48-d65a-11e6-8d26-af049c26c330.jpg)

**Comment by gabriele-v** [comment]
Agree with Nikolay for new databases, we need to modify script in some way to identify field that needs translation. 

Regarding fix existing DB, why not simply release a DB upgrade version with something like this (pseudo-SQL):
UPDATE Currency SET Symbol = "€" WHERE Code = "EUR"

**Comment by stef145g** [comment]
Checking further: the problem is not with python because the resulting DB_Table_CURRENCYFORMATS_V1 file is encoded as utf-8

The  resulting code is : 
```
wxString euro = wxString::Format("INSERT INTO CURRENCYFORMATS_V1 VALUES ('2', '%s', '€', '', '.', ' ', '', '', '100', '1', '%s')", wxTRANSLATE("European euro"), wxTRANSLATE("EUR"));
db->ExecuteUpdate(euro);
```
I have separated the command in two to see the result of 'euro'

Using the same line of code where I define *euro* to see the resulting string, the utf-8 character is OK when added to the call before calling the .h code.
When I add the same line of code in the .h file, the result is corrupted.
I attempted to put the code in a .cpp file for DB_Table_CURRENCYFORMATS_V1.h but get the same error.

I suspect it has something to do with how utf-8 code is used in wxWidgets. There is a call where the code is passing a wxString but expects a const *char and the conversion is getting corrupted within wxWidgets.

**Comment by stef145g** [comment]
@gabriele-v Your method would be OK when the database has all its currencies set. Some of the old databases are more than likely, still being used (Mine being a classic example) that will not have all the currencies installed.

Also any new database created will have the same problem, unless after being created on the 2nd run the database is immediately upgraded.

**Comment by vomikan** [comment]
@stef145g I've provided the working file ( [see above](https://github.com/moneymanagerex/moneymanagerex/files/687910/DB_Table_Currencyformats_V1.h.txt) ) It's utf8 but with BOM. Current file encoding is  utf8 without BOM.
Also (I do not get tired of repeating) #pragma execution_character_set("utf-8") recomended.
And Py script generate too many useless wxTRANSLATION

**Comment by gabriele-v** [comment]
@stef145g 
We should provide 2 fixes, one for new created DB and one for existing ones:
1. For new databases => Working on C++ code side as discussed by you and Nikolay
2. Create DB Version 8 that will patch currencies basing on Currency Symbol ( = ISO code)

I know that second won't fix all DBs (older ones don't have ISO code populated) but this issue seems to occur only after we have inserted translations into DB_Table_CURRENCYFORMATS_V1.h, so every DB created in this way has ISO codes too. Older DB will still remains without all currencies, but at least we can solve problem on recently created DB.

**Comment by stef145g** [comment]
Taking up @vomikan 's suggestion, sqlite2cpp.py has been modified to provide the correct output for building the .h files.

The database definitions in table_v1.sql for translations have been modified. Data requiring translation require a `_tr_` prefix to the name. This allows sqlite2cpp to correctly identify the fields requiring translation and is treated accordingly.

**Comment by vomikan** [comment]
Now it working as expected.
This code is exactly what I've mentioned early.
![screenshot 3](https://cloud.githubusercontent.com/assets/6836805/21980666/36408f54-dbf5-11e6-8991-7081345c12c2.jpg)
Thank you @stef145g ! Exelent job.
